File: packages/typescript/core-sdk/src/pagination.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { paginate, collectAll, type Page } from "./pagination.js";

test("paginate walks every page in order", async () => {
  const pages: Record<string, Page<number>> = {
    "": { items: [1, 2], nextCursor: "p2" },
    p2: { items: [3, 4], nextCursor: "p3" },
    p3: { items: [5], nextCursor: undefined },
  };
  const fetch = async (cursor?: string) => pages[cursor ?? ""];

  const got: number[] = [];
  for await (const item of paginate(fetch)) {
    got.push(item);
  }
  assert.deepEqual(got, [1, 2, 3, 4, 5]);
});

test("collectAll stops at an empty cursor", async () => {
  let calls = 0;
  const fetch = async () => {
    calls++;
    return { items: ["only"], nextCursor: undefined } as Page<string>;
  };
  const items = await collectAll(fetch);
  assert.deepEqual(items, ["only"]);
  assert.equal(calls, 1);
});

File: packages/flutter/core_sdk/test/pagination_test.dart
import 'package:test/test.dart';
import 'package:core_sdk/core_sdk.dart';

void main() {
  test('paginate walks every page in order', () async {
    final pages = <String, Page<int>>{
      '': Page([1, 2], 'p2'),
      'p2': Page([3, 4], 'p3'),
      'p3': Page([5], null),
    };
    Future<Page<int>> fetch(String? cursor) async => pages[cursor ?? '']!;

    final got = await paginate(fetch).toList();
    expect(got, [1, 2, 3, 4, 5]);
  });

  test('collectAll stops at a null cursor', () async {
    var calls = 0;
    Future<Page<String>> fetch(String? cursor) async {
      calls++;
      return const Page(['only'], null);
    }

    final items = await collectAll(fetch);
    expect(items, ['only']);
    expect(calls, 1);
  });
}
